proposals/NNNN-offsetting-subscript.md

```diff
@@ -0,0 +1,332 @@
+```
+let s = "Hello, Swift!"
+let subject = s[s.index(s.startIndex, offsetBy: 7)...s.index(s.startIndex, offsetBy: 11)]
+```
+This approach, unfortunately, suffers from redundancy and is in general unwieldy 
+to handle. 
+
+A shorter approach, which is also available, is to use combinations of `prefix`, 
+`suffix` and the `drop` variants. A solution using these would follow like such:
+```
+let subject = s.suffix(6).prefix(5)
+```
+While this is much shorter, it suffers from multiple drawbacks. It is not as
+natural as using a range, which increases the cognitive load for a user. This 
+solution also suffers from API discoverability issues, since a user must learn 
+multiple methods and figure out that they can be composed in this way. 
+
+## Proposed solution
+A solution we propose to this problem is to extend `Collection` and 
+`RangeReplaceableCollection` with a subscript that takes a range, which 
+would be used to offset an index of a collection.
+`Collection` and `MutableCollection` will receive subscripts that take a
+single `Int` to return a single element from the collection.
+
+A highly request ability for getting a slice of a collection is the ability to
+offset relative to the `endIndex` of a collection. Currently range support can
+cover most of the desired requests, however, due to ranges requiring an 
+upperBound >= lowerBound constraint certain use cases are not met. To solve this
+Four new operators should be added along with a new type to help model this 
+behavior. To encapsulate this behavior, a new protocol should also be added.
+The other range types will then also conditionally conform to this protocol.
+
+## Detailed design
+A new protocol Should be added:
+```
+/// A type that can be used to get a slice of a collection relative to the
+/// collections start and end indices.
+public protocol RelativeOffsetRangeExpression {
+  /// A type that represents an offset range to slice a collection relatively.
+  func relativeOffsetRange() -> RelativeOffsetRange
+}
+```
+A new type to represent a range offset:
+```
+/// Returns a relative offset range that represents a slice of a collection
+/// given two offset values relative to the collections start and/or end 
+/// indices.
+///
+/// You can use the RelativeOffsetRange to represent a slice of a collection 
+/// relative with offset values relative the collections start/end indices.
+/// If the offset value is greater or equal to 0 it will be relative to the 
+/// `startIndex`, else it will be relative to the `endIndex`. For example:
+///
+///     let exceptLastFour = ..<(-4)
+///
+///     let r1 = exceptLastFour.relativeAsOffset()
+///     // r1 == 0..<-4
+///
+/// The `r1` range is bounded on the lower end by `0` because that is the
+/// offset that does not change `startIndex`. In the next example we'll see
+/// how a partial range from works:
+///
+///     let greeting = "Hello, Swift!"
+///     let greeting = numbers[offset: (-6)...]
```

Comment:
`numbers`? This should probably be `s`

Reply:
It should be let name = greeting. Thanks